### PR TITLE
fix!: pin dependency version in all wrapper conda env yamls

### PR DESCRIPTION
**Pull request by tedil**
This PR aims to increase reproducibility and is a precursor to be able to properly end-to-end test the pipeline.
All version specifications in the wrappers' environment.yaml files were pinned to one specific version, in a configuration that can actually be installed with mamba/conda and a strict channel priority (which is why sometimes the channel definitions were also changed).
This was done programmatically, with the exception of some wrappers for which manual intervention was required:
 - gcnv
 - jaffa/run
 - cnvetti/on_target/postprocess
 - svtk
 - scramble
 - link_in_bam
 - lohhla

Since there are no tests for the wrappers yet, I can't really be sure if this breaks anything, so I'd rather be cautious and assume that it does.

### CODE REVIEW

File: snappy_wrappers/wrappers/bed_jaccard_operations/environment.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/bed_jaccard_operations/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/canvas/somatic_wgs/environment.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/canvas/somatic_wgs/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/cbioportal/generate_cna/environment.yaml [hunk 3]
```diff
@@ -1,14 +1,14 @@
 name: generate_cna
 
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/cbioportal/generate_cna/environment.yaml:6]: remove

File: snappy_wrappers/wrappers/cbioportal/merge_tables/environment.yaml [hunk 4]
```diff
@@ -1,14 +1,14 @@
 name: merge_tables
 
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/cbioportal/merge_tables/environment.yaml:6]: remove

File: snappy_wrappers/wrappers/cnvetti/on_target/coverage/environment.yaml [hunk 5]
```diff
@@ -1,8 +1,8 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/cnvetti/on_target/coverage/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/cnvetti/on_target/postprocess/environment.yaml [hunk 6]
```diff
@@ -1,20 +1,20 @@
 channels:
-- conda-forge
-- bioconda
-- anaconda
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/cnvetti/on_target/postprocess/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/cnvetti/on_target/postprocess/environment.yaml [hunk 7]
```diff
@@ -1,20 +1,20 @@
...
-- bioconductor-dnacopy
-- bioconductor-cghcall
-- bioconductor-rsamtools
-- bioconductor-org.hs.eg.db
-- bioconductor-txdb.hsapiens.ucsc.hg19.knowngene
-- bcftools ==1.5
-- samtools ==1.5
-- htslib ==1.5
-- openblas
+  - r==4.2
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/cnvetti/on_target/postprocess/environment.yaml:6]: should be r-base

File: snappy_wrappers/wrappers/cnvetti/wgs/environment.yaml [hunk 8]
```diff
@@ -1,8 +1,8 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/cnvetti/wgs/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/cnvkit/environment.yaml [hunk 9]
```diff
@@ -1,8 +1,8 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/cnvkit/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/cnvkit/postprocess/environment.yaml [hunk 10]
```diff
@@ -1,8 +1,8 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/cnvkit/postprocess/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/control_freec/environment.yaml [hunk 11]
```diff
@@ -1,8 +1,8 @@
 channels:
-- conda-forge
-- bioconda
-- base
+  - conda-forge
+  - bioconda
+  - base
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/control_freec/environment.yaml:4]: not sure this exists?

File: snappy_wrappers/wrappers/copywriter/environment.yaml [hunk 12]
```diff
@@ -1,23 +1,23 @@
...
-- bioconductor-org.hs.eg.db
-- bioconductor-ensdb.hsapiens.v75
-- bioconductor-ensdb.hsapiens.v86
-- bioconductor-txdb.hsapiens.ucsc.hg19.knownGene
-- bioconductor-txdb.hsapiens.ucsc.hg38.knownGene
-- bcftools
-- samtools
-- htslib
-- openblas
+  - r==3.6
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/copywriter/environment.yaml:6]: should be r-base

File: snappy_wrappers/wrappers/delly2/germline/environment.yaml [hunk 13]
```diff
@@ -1,8 +1,8 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/delly2/germline/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/dupradar/environment.yaml [hunk 14]
```diff
@@ -1,7 +1,7 @@
 channels:
-- conda-forge
-- bioconda
+  - conda-forge
+  - bioconda
 dependencies:
-- r
-- bioconductor-dupradar
-- samtools
+  - r==4.3
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/dupradar/environment.yaml:5]: r-base

File: snappy_wrappers/wrappers/epitoper/environment.yaml [hunk 15]
```diff
@@ -1,9 +1,9 @@
 channels:
-- conda-forge
-- bioconda
-- r
-- http://conda.cubi.bihealth.org/cubiconda/
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/epitoper/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/epitoper_par/environment.yaml [hunk 16]
```diff
@@ -1,6 +1,6 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/epitoper_par/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/fusioncatcher/run/environment.yaml [hunk 17]
```diff
@@ -1,7 +1,7 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/fusioncatcher/run/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/gatk_hc_par/environment.yaml [hunk 18]
```diff
@@ -1,12 +1,12 @@
 channels:
-- conda-forge
-- bioconda
-- r
-- http://conda.cubi.bihealth.org/cubiconda
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/gatk_hc_par/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/gatk_read_backed_phasing/environment.yaml [hunk 19]
```diff
@@ -1,11 +1,11 @@
 channels:
-- conda-forge
-- bioconda
-- r
-- http://conda.cubi.bihealth.org/cubiconda
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/gatk_read_backed_phasing/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/gatk_read_backed_phasing_par/environment.yaml [hunk 20]
```diff
@@ -1,11 +1,11 @@
 channels:
-- conda-forge
-- bioconda
-- r
-- http://conda.cubi.bihealth.org/cubiconda
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/gatk_read_backed_phasing_par/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/gatk_ug_par/environment.yaml [hunk 21]
```diff
@@ -1,11 +1,11 @@
 channels:
-- conda-forge
-- bioconda
-- r
-- http://conda.cubi.bihealth.org/cubiconda
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/gatk_ug_par/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/gcnv/environment.yaml [hunk 22]
```diff
@@ -1,42 +1,40 @@
 channels:
   - conda-forge
   - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/gcnv/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/hts_screen/environment.yaml [hunk 23]
```diff
@@ -1,9 +1,9 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/hts_screen/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/jaffa/run/environment.yaml [hunk 24]
```diff
@@ -1,7 +1,8 @@
 channels:
 - conda-forge
 - bioconda
+- r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/jaffa/run/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/jannovar/environment.yaml [hunk 25]
```diff
@@ -1,12 +1,11 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/jannovar/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/manta/germline_wgs/environment.yaml [hunk 26]
```diff
@@ -1,7 +1,7 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/manta/germline_wgs/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/manta/somatic_wgs/environment.yaml [hunk 27]
```diff
@@ -1,6 +1,6 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/manta/somatic_wgs/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/minialign/run/environment.yaml [hunk 28]
```diff
@@ -1,10 +1,10 @@
 channels:
-- conda-forge
-- bioconda
-- r
-- http://conda.cubi.bihealth.org/cubiconda
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/minialign/run/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/mutect/environment.yaml [hunk 29]
```diff
@@ -1,9 +1,9 @@
 channels:
-- conda-forge
-- bioconda
-- r
-- http://conda.cubi.bihealth.org/cubiconda
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/mutect/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/popdel/environment.yaml [hunk 30]
```diff
@@ -1,9 +1,9 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/popdel/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/rnaqc/dupradar/environment.yaml [hunk 31]
```diff
@@ -1,8 +1,8 @@
...
-- bioconda
+  - conda-forge
+  - bioconda
 dependencies:
-- samtools
-- r
-- bioconductor-dupradar
-- freetype=2.9.1
+  - samtools==1.12
+  - r==4.0
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/rnaqc/dupradar/environment.yaml:6]: r-base

File: snappy_wrappers/wrappers/rnaqc/environment.yaml [hunk 32]
```diff
@@ -1,11 +1,11 @@
...
+  - bioconda
 dependencies:
-- r
-- bioconductor-dupradar
-- samtools
-- rna-seqc ==1.1.8
-- openjdk=7
-- rseqc
-- freetype=2.9.1
+  - r==4.0
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/rnaqc/environment.yaml:5]: r-base

File: snappy_wrappers/wrappers/scalpel/somatic/environment.yaml [hunk 33]
```diff
@@ -1,8 +1,8 @@
 channels:
-- conda-forge
-- bioconda
-- r
+  - conda-forge
+  - bioconda
+  - r
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/scalpel/somatic/environment.yaml:4]: remove

File: snappy_wrappers/wrappers/signatures/deconstruct_sigs/environment.yaml [hunk 34]
```diff
@@ -1,10 +1,9 @@
...
+  - conda-forge
+  - bioconda
 dependencies:
-- r #==3.2.2
-- r-tidyverse
-- r-deconstructsigs
-- bioconductor-bsgenome.hsapiens.ucsc.hg19
-- openblas
-
+  - r==4.3
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/signatures/deconstruct_sigs/environment.yaml:5]: r-base

**Review by Nicolai-vKuegelgen — CHANGES_REQUESTED** [review]
From my understanding the 'r' channel should not be used anymore, also some packages use 'r' as a package name instead of r-base.

I don't think "http://conda.cubi.bihealth.org/cubiconda" is reachable from the outside, so relying on that here is a bit unfortunate (no easy way to fix this though, I guess)

File: snappy_wrappers/wrappers/gatk4_hc/environment.yaml [hunk 35]
```diff
@@ -2,9 +2,9 @@ channels:
...
-  - bcftools =1.16
-  - samtools =1.16
-  - htslib =1.16
-  - parallel =20221122
-  - picard-slim =2.27
+  - gatk4==4.5.0.0
+  - bcftools==1.16
+  - samtools==1.16.1
+  - htslib==1.16
+  - parallel==20221122
```
Comment (Nicolai-vKuegelgen) [inline snappy_wrappers/wrappers/gatk4_hc/environment.yaml:9]: I have pinned version 20231122 in a local branch, I think that might better here

**Review by Nicolai-vKuegelgen — APPROVED** [review]
LGTM overall

However, this covers many more steps than I am familiar with, so can't guarantee that this won't break anything.

File: snappy_wrappers/wrappers/scarHRD/environment.yaml
Comment (ericblanc20) [inline snappy_wrappers/wrappers/scarHRD/environment.yaml:1]: Is it possible to keep the comments here?

File: snappy_wrappers/wrappers/sequenza/install/environment.yaml
Comment (ericblanc20) [inline snappy_wrappers/wrappers/sequenza/install/environment.yaml:1]: Is it possible to keep the comments here?

**Review by ericblanc20 — COMMENTED** [review]
Looks good to me

File: snappy_wrappers/wrappers/scarHRD/environment.yaml
Comment (tedil) [reply snappy_wrappers/wrappers/scarHRD/environment.yaml:1]: Yes, good point!

**Pull request merged**